src/libsyntax/mut_visit.rs

```diff
@@ -586,15 +585,24 @@ pub fn noop_visit_tts<T: MutVisitor>(TokenStream(tts): &mut TokenStream, vis: &m
     })
 }
 
-// apply ident visitor if it's an ident, apply other visits to interpolated nodes
-pub fn noop_visit_token<T: MutVisitor>(t: &mut TokenKind, vis: &mut T) {
-    match t {
+// Apply ident visitor if it's an ident, apply other visits to interpolated nodes.
+// In practice the ident part is not actually used by specific visitors right now,
+// but there's a test below checking that it works.
+pub fn noop_visit_token<T: MutVisitor>(t: &mut Token, vis: &mut T) {
+    let Token { kind, span } = t;
+    match kind {
+        token::Ident(name, _) | token::Lifetime(name) => {
+            let mut ident = Ident::new(*name, *span);
+            vis.visit_ident(&mut ident);
+            *name = ident.name;
```

Comment:
shouldn't this write back the `span`, too?

Reply:
There's no `span` to write back to (`token::Ident` no longer have a span).
The span is now in `Token` only and it should be visited by `vis.visit_span(span)` a few lines below.

Reply:
well, you can write back to the `span` of the `Token`. If the implementor of `visit_ident` modifies the `Span` in `ident`, then that modification will get lost here. The fact that `visit_span` is visited below does not change this. The logic for modifying spans may theoretically only be implemented in `visit_ident` and depend on the `name`.